### PR TITLE
Average pool CPU with python code

### PR DESCRIPTION
**Pull request by nouiz**
I added some follow up in gh-2792
fix gh-2774

### CODE REVIEW

**Comment by nouiz** [comment]
This PR is ready for review. I have a follow ticket: https://github.com/Theano/Theano/issues/2792

File: theano/sandbox/cuda/tests/test_dnn.py [hunk 1]
```diff
@@ -179,11 +177,11 @@ def fn(x):
                         for node in fg.maker.fgraph.toposort()])
             g_out = fg(data)
 
-            if func is T.max:
+            if True:
```
Comment (abergeron) [inline theano/sandbox/cuda/tests/test_dnn.py]: I know you do this to avoid indentation changes, but I still don't like it.

File: theano/tensor/signal/downsample.py [hunk 2]
```diff
@@ -62,11 +63,16 @@ def max_pool_2d(input, ds, ignore_border=False, st=None, padding=(0, 0)):
             of the images, pad_h is the size of the top and bottom margins,
             and pad_w is the size of the left and right margins.
     :type padding: tuple of two ints
+    :param mode: 'max', 'average_inc_pad' or 'average_exc_pad',
+        the operation executed on each window. The max always exclude the
+        padding in the computation of the max. For average, you have the choise.
```
Comment (abergeron) [inline theano/tensor/signal/downsample.py]: I would rewrite the doc as:
```
Operation executed on each window.  `max` always excludes the padding in the computation.  `average` gives you the choice to include or exclude it.

File: theano/tensor/signal/downsample.py [hunk 3]
```diff
@@ -178,9 +185,9 @@ def out_shape(imgshape, ds, ignore_border=False, st=None, padding=(0, 0)):
         rval = list(imgshape[:-2]) + [nr, nc]
         return rval
 
-    def __init__(self, ds, ignore_border=False, st=None, padding=(0, 0)):
-        """
-        :param ds: downsample factor over rows and column.
+    def __init__(self, ds, ignore_border=False, st=None, padding=(0, 0),
+                 mode='max'):
+        """:param ds: downsample factor over rows and column.
```
Comment (abergeron) [inline theano/tensor/signal/downsample.py]: Why do you put the `:param:` list on the very first line?

File: theano/tensor/signal/downsample.py [hunk 4]
```diff
@@ -217,11 +229,16 @@ def __init__(self, ds, ignore_border=False, st=None, padding=(0, 0)):
...
+                "DownsampleFactorMax mode parameter only support 'max',"
+                " 'average_inc_pad' and 'average_exc_pad'. Got %s" % mode)
+        self.mode = mode
 
     def __str__(self):
-        return '%s{%s, %s, %s, %s}' % (
+        return '%s{%s, %s, %s, %s, %s}' % (
             self.__class__.__name__,
-            self.ds, self.st, self.ignore_border, self.padding)
+            self.ds, self.st, self.ignore_border, self.padding, self.mode)
```
Comment (abergeron) [inline theano/tensor/signal/downsample.py]: This *might* be better served by the default \__props__ implementation.

File: theano/tensor/signal/downsample.py [hunk 5]
```diff
@@ -251,27 +268,37 @@ def perform(self, node, inp, out):
...
                 (x.shape[0], x.shape[1], img_rows, img_cols),
-                dtype=x.dtype) + fill
+                dtype=x.dtype)
             y[:, :, pad_h:(img_rows-pad_h), pad_w:(img_cols-pad_w)] = x
         else:
             y = x
+        func = numpy.max
+        if self.mode != 'max':
+            func = numpy.average
         # max pooling
```
Comment (abergeron) [inline theano/tensor/signal/downsample.py]: This comment is no longer always true

File: theano/sandbox/cuda/tests/test_dnn.py (hunk 1 above)
Comment (nouiz) [inline theano/sandbox/cuda/tests/test_dnn.py]: Yes. It make the diff more clear, but I agree, I should make a separate commit that just do the indentation.

File: theano/tensor/signal/downsample.py (hunk 3 above)
Comment (nouiz) [inline theano/tensor/signal/downsample.py]: probably auto indentation did that. It expect a first line of description.

**Comment by nouiz** [comment]
I fixed all comments + one other docstring update.
